Light and dark theme

keller-mark opened the pull request:
I implemented light/dark themes by adding a `theme` property to the api.js config schema. This PR moves most scss color styles out of the existing scss files and into `src/css/themes/light.scss` and `src/css/themes/dark.scss` (with light values in the light file and the existing values in the dark file).

This PR adds a theme class to the top-level `vitessce-container` div (`<div className="vitessce-container vitessce-theme-dark">`) which determines the styles inherited by the children.
I think there are tons of approaches (and npm packages) one could take to implement themes so if you had a different approach in mind I can close this and reconsider, this was simply the first thing that came to mind.

Fixes #405

mccalluc commented on src/app/Welcome.js:
```diff
@@ -93,7 +93,7 @@ function Info() {
   const { configs } = props;
   return (
-    <div className="vitessce-container">
+    <div className="vitessce-container vitessce-theme-dark">
```
Should this also be pulled from api.js? If Nils decides tomorrow he wants the light theme for the whole demo, can that be a one line change?

(Or maybe not: The priority is the Vitessce component... Sorting out the style rules that are used only in Welcome so they can be either dark or light would be a wasted effort if we never want the whole demo to be light. If this is intentional, add a comment explaining?)

mccalluc replied:
Ignore that: api.js will be moving to vitessce-data, and this display detail should stay here, and redeploy is fine.

mccalluc commented on src/css/themes/light.scss:
```diff
@@ -0,0 +1,68 @@
+        border: 1px solid #555;
+        color: #000;
+    }
+}
```
I'd worry about keeping these two in sync. Could you make this a single [mixin with an argument](https://sass-lang.com/documentation/at-rules/mixin#arguments)?

keller-mark replied:
Great idea, I did not know about that feature

mccalluc commented on src/schemas/dataset.schema.json:
```diff
@@ -75,6 +75,7 @@
     },
-    "staticLayout": { "$ref": "#/definitions/components" }
+    "staticLayout": { "$ref": "#/definitions/components" },
+    "theme": { "type": "string" }
```
```suggestion
    "theme": {
      "type": "string",
      "enum": ["dark", "light"]
     }
```
or something like that.

mccalluc commented on src/css/index.scss:
```diff
@@ -17,6 +20,14 @@
+  }
+  &.vitessce-theme-light {
+    @include theme-light;
+  }
```
```suggestion
  &.vitessce-theme-dark {
    @include vitessce-theme("dark");
  }
  &.vitessce-theme-light {
    @include vitessce-theme("light");
  }
```

keller-mark commented:
I have updated this PR to use sass maps rather than if-else-if statements. Now if we want to add a theme, we just need to add a new color map object here https://github.com/hubmapconsortium/vitessce/pull/494/files#diff-11c77daa14f0b2a567413fc3942dd8a4R1

I wrapped all the scss mixins that depend on the theme parameters in their own `vitessce-themable` mixin here https://github.com/hubmapconsortium/vitessce/pull/494/files#diff-b3e069d412525c3573780a028b971af9R12

I also added logic to obtain the light/dark theme preference from the browser / OS https://github.com/hubmapconsortium/vitessce/pull/494/files#diff-c948cf7605859408caf11c4ca94757ebR79

![Screen Shot 2020-03-16 at 3 24 19 PM](https://user-images.githubusercontent.com/7525285/76792898-62e4ae80-679a-11ea-83e5-2bbcb12e9931.png)


Screenshots:

- light

<img width="1440" alt="Screen Shot 2020-03-16 at 3 24 07 PM" src="https://user-images.githubusercontent.com/7525285/76792913-6b3ce980-679a-11ea-9371-346c81dd0534.png">

<img width="1440" alt="Screen Shot 2020-03-16 at 3 20 44 PM" src="https://user-images.githubusercontent.com/7525285/76792920-7132ca80-679a-11ea-9b59-948ab1dbe8bd.png">

- dark

<img width="1439" alt="Screen Shot 2020-03-16 at 3 23 59 PM" src="https://user-images.githubusercontent.com/7525285/76792933-7728ab80-679a-11ea-9174-03692d96c437.png">

<img width="1440" alt="Screen Shot 2020-03-16 at 3 23 50 PM" src="https://user-images.githubusercontent.com/7525285/76792943-7abc3280-679a-11ea-8505-3dd2f1890086.png">

mccalluc commented on src/app/PubSubVitessceGrid.js:
```diff
@@ -2,6 +2,7 @@ import React from 'react';
 import VitessceGrid from 'vitessce-grid';
 
 import { LayerPublisher } from '../components/layerpublisher';
+import { getPreferredTheme } from '../components/utils';
```
This is nifty, but I have mixed feelings about it. I feel like projects generally have a solid sense of their own style, and having isolated components swapping their contrast for some users because of a browser setting would be unexpected -- I think that if this is used, it should be at a higher level and passed down.

Do you have a sense of how likely users are to twiddle this in their browser? Is it primarily an accessibility setting?

Down below you do let it be overridden by the config, but I think that's going to be tied more closely to the data... I think either a boolean or a theme-name component prop may be better.

keller-mark replied:
Very true. I will get rid of the default-to-user-preference stuff. Instead, how do you feel about the following:
- default to dark theme
- show toggle on welcome page to change the user preference between light/dark
    - save this to localstorage?
- dataset config `theme` property overrides both the default and the toggle setting
- url parameter overrides all three (default dark, toggle setting, theme property)

mccalluc replied:
I probably wouldn't cascade through so many options, but it's subjective. I'd be happy with a react prop that is set from a url parameter, and I'm not sure we have a story than requires anything more complicated than that. The code here is just a demo, and if the downstream user want to use localstorage, or put an extra setting in the config, they are free to do so... I'm just not sure we want to put cost of maintaining those options in this library.

... So, to wrap up, I think it'd be better to keep it simpler, but if you've got a compelling use case in mind, just go for it: No need for a long thread here. Thanks!

mccalluc commented on src/css/_popover.scss:
```diff
@@ -24,7 +23,7 @@
                 }
                 &:not(:last-child) {
-                    border-bottom: 1px solid $color-gray;
+                    border-bottom: 1px solid #808080;
```
I know this is right in the middle... but if it were changed, it would need to be changed in two ways... so I'd use a mag-get here, just as a reminder, even if it is degenerate.

mccalluc commented on src/css/_sets-manager.scss:
```diff
@@ -87,12 +88,13 @@
                     does not take padding into account. */
                 width: calc(100% - 3.75rem);
                 position: fixed;
+                border-bottom: 1px solid #555555;
```
Here and below, if colors wouldn't change, and it's intentional, maybe add a note so no one worries later?

mccalluc commented:
@keller-mark : Replied to you above. I'd prefer to keep it as simple as possible, but I'll trust your judgement to wrap this up... but feel free to ping me if you want more opinions!

mccalluc commented on src/components/utils.js:
```diff
@@ -71,3 +71,15 @@ export function fromEntries(iterable) {
+    return els[0];
+  }
+  return document.body;
+}
```
I'm not sure how this is used, but I'm pretty sure Vitessce should never be messing with the document root. I don't want #459 to be undone. Sorry if I'm not getting it.

keller-mark replied:
That makes sense I will think about other approaches to use here.

The issue is that the Popover element by default is rendered as a child of `document.body` (see `getPopupContainer` description [here](https://ant.design/components/tooltip/#API)) but the theme styles are applied by adding a theme class to the `.vitessce-container` div. So when the popover element is outside that div, the theme cannot be accessed. 

A different approach I can try is to render an empty div element as a sibling of the Popover component, use a React `ref` to get the DOM node object for the empty div, and then pass that DOM node to `getPopupContainer`

keller-mark replied:
<img width="933" alt="Screen Shot 2020-03-20 at 11 39 21 PM" src="https://user-images.githubusercontent.com/7525285/77218758-1b954f80-6b05-11ea-8c02-459d839f67b7.png">

mccalluc replied:
Cool: Either way, I think it might be a little confusing to someone down the road, so may just add the explanation above in a comment, and use whichever approach seems best to you. I'm not sure myself. Thanks!
